libs/utils/android/screen.py

```diff
@@ -155,4 +157,70 @@ def unlock(target):
        System.menu(target)
        System.home(target)
 
+    class Recorder(object):
+        """
+        Android screenrecord utility class
+        """
+
+        def __init__(self, target):
+            self._target = target
+            self._log = logging.getLogger('Recorder')
+
+            self._recording = False
+
+        def start(self, out_file,
+                         timestamp=True, max_time=None):
+            """
+            Start recording the screen of the device
+            """
+            if self._recording:
+                raise RuntimeError('Recording already started')
+
+            self._local_file = out_file
+            self._remote_file = self._target.path.join(
+                self._target.working_directory,
+                os.path.basename(out_file)
+            )
+
+            self._max_time = max_time
+
+            options = []
+            if timestamp:
+                options.append('--bugreport')
+            if max_time:
+                options.append('--time-limit {}'.format(max_time))
+
+            record_cmd = 'screenrecord {} {}'.format(
+                ' '.join(options), self._remote_file
+            )
+
+            self._log.debug('screenrecord command = \"{}\"'.format(record_cmd))
+            self._log.info('Starting screen recording')
+
+            self._recording = True
+            self._process = self._target.background(record_cmd, as_root=True)
+
+        def stop(self):
+            """
+            Stop recording the screen of the device
+            """
+            if not self._recording:
+                raise RuntimeError('Recording not started')
+
+            if not self._max_time:
+                # Recording must be ended by Ctrl-C on the device
+                self._log.debug('Sending Ctrl-C to screenrecord')
+                self._target.execute('kill -INT $(pgrep screenrecord)', as_root=True)
+
+            self._log.info('Waiting for recording to be completed')
+            self._process.wait()
```

Comment:
Ok, sorry, I was unclear. Let me clarify:

> we're waiting for the SIGINT to be processed

Indeed, but that _should_ be quick, and it's part of a normal `stop()` procedure. When I was saying "we should really wait" I meant that we wait for the recorder to timeout, which can be very long (potentially).

I see two different semantics that should be decorrelated. `Record` should provide 3 methods: `Recorder.start()`, `Recorder.stop()` and `Recorder.wait()`. `Recorder.stop()` should end the recorder immediately, meaning that it will issue a SIGINT unilaterally, without caring about `max_time`,  and use `self._process.wait()` to do that cleanly. On the other hand, `Recorder.wait()` should do nothing else but wait for `max_time` to elapse. That way, we can either:

- use `stop()` to stop the recorder as soon as a workload completes for example (in which case `max_time` will just serve as a timeout ensuring that  we don't record forever if the workload bugs and never ends)
- use `wait()` to record the screen for a specific amount of time (specified by `max_time`).

What do you think ? I guess you already have all the stuff you need to implement that, it's just a bit of refactoring :smiley: